### PR TITLE
Counting the opened sockets

### PR DESCRIPTION
**Pull request by GuillaumeGaudin**
Following the number of opened socket in order to know the number of users. 
Allow to limit access to specific service for a wanted number of user.

### CODE REVIEW

File: core/router/src/main/java/org/wisdom/router/WebSocketRouter.java [hunk 1]
```diff
@@ -388,4 +385,39 @@ public void send(String uri, String client, byte[] message) {
...
+
+    /**
+     * Get the number of opened socket for each uri counting when dispatchers uses a same Uri
+     *
+     * @return map of uri with the number of opened sockets
+     */
+    @Override
+    public Map<String, Integer> getNumberOpenedSocketsByUri() {
+        Map<String, Integer> numberOpenedSocketByUri = new LinkedHashMap<>();
+        for (WebSocketDispatcher dispatcher : dispatchers) {
```
Comment (cescoffier) [inline core/router/src/main/java/org/wisdom/router/WebSocketRouter.java]: it needs to be protected against concurrent modification

File: core/router/src/main/java/org/wisdom/router/WebSocketRouter.java [hunk 2]
```diff
@@ -388,4 +385,39 @@ public void send(String uri, String client, byte[] message) {
...
+    /**
+     * Get the number of opened socket for a specific uri
+     *
+     * @param uri   the uri of which the opened sockets are counted
+     * @return the number of opened socket for this uri
+     */
+    @Override
+    public Integer getUriOpenedSockets(String uri) {
+        Integer numberOpenedSockets = 0;
+        for (WebSocketDispatcher dispatcher : dispatchers) {
```
Comment (cescoffier) [inline core/router/src/main/java/org/wisdom/router/WebSocketRouter.java]: need to be protected against concurrent accesses

File: core/wisdom-vertx-engine/src/main/java/org/wisdom/framework/vertx/WisdomVertxServer.java [hunk 3]
```diff
@@ -450,4 +450,33 @@ public void received(String uri, byte[] content, Socket socket) {
...
+
+    /**
+     * Get map of uri and number of opened sockets
+     *
+     * @return map of uri with the number of opened sockets
+     */
+    @Override
+    public Map<String, Integer> getNumberOpenedSockets() {
+        Map<String, Integer> numberOpenedSocketByUri = new LinkedHashMap<>();
+        for (Map.Entry<String, List<Socket>> socketByUri : socketsByUri.entrySet()) {
```
Comment (cescoffier) [inline core/wisdom-vertx-engine/src/main/java/org/wisdom/framework/vertx/WisdomVertxServer.java]: must be protected against concurrent accesses

File: core/wisdom-vertx-engine/src/main/java/org/wisdom/framework/vertx/WisdomVertxServer.java [hunk 4]
```diff
@@ -450,4 +450,33 @@ public void received(String uri, byte[] content, Socket socket) {
...
+    /**
+     * Get the number of opened socket for a specific uri
+     *
+     * @param uri
+     * @return the number of opened socket for this uri
+     */
+    @Override
+    public Integer getNumberOpenedSocketsByUri(String uri) {
+        Integer numberOpenedSocketForUri = 0;
+        if (socketsByUri.containsKey(uri)) {
```
Comment (cescoffier) [inline core/wisdom-vertx-engine/src/main/java/org/wisdom/framework/vertx/WisdomVertxServer.java]: concurrency issue too

**Comment by cescoffier** [comment]
Can you add some tests? In addition, it needs to be protected against concurrent accesses.

Finally, please use English in your tickets.

**Comment by cescoffier** [comment]
Thanks!

**Comment by jetien** [comment]
Good job ! Hope you're fine Clément

**Pull request merged**